text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+
+### No custom parsing or error-handling
+Extension functions provide more full-featured constructors through custom parsing and error handling, but Cedar functions provide no such facilities. This may make them harder to read, write, and understand.
+
+For example, you could encode a decimal number as a `Long`, and then make Cedar functions to construct and compare decimals:
+```
+function mydecimal(i,f) {
+  if f >= 0 && f <= 9999 then
+    i * 10000 + f    // will overflow if i too big
+  else
+    18446744073709551615 + 1    // fraction f too big: induce overflow
+};
+function mydecimalLTE(d,e) {
+  d <= e    // d and e are just Long numbers
+}
+```
+These functions basically implement the equivalent of Cedar `decimal` numbers. But the approach has at least two drawbacks.
+
+First, if `i` and/or `f` are outside the allowed range, you will get a Cedar overflow exception at run-time. This exception is not as illuminating as the custom error emitted by the `decimal` extension function (whose message will be `"Too many digits"`).
+Moreover, custom errors from constructor parameter validity checks can be emitted during validation when using extension functions, but not when using Cedar functions.
+
+Second, there is no special parsing for Cedar functions. With Cedar's built-in `decimal`, you can write `decimal("123.12")` which more directly conveys the number being represented than does `mydecimal(123,12)`.
+
+Of course, these drawbacks do not necessarily speak against Cedar functions generally, but suggest that for suitably general use-cases (like decimal numbers!), an extension function might be warranted instead.
+
+### Readability: Policies are no longer standalone
+A policy can no longer be read by itself, it has to be read in the context of all function definitions it uses.
+Policies that use a large number of functions may be hard to read.
+
+## Alternatives
+
+### Type annotations
+We could require/allow Cedar functions to have type annotations, taking type definitions from either the schema or allowing them to be inline.
+
+Example:
+
+Schema:
+```
+type SemVer = { major : Long, minor : Long, patch : Long };
+```
+Policy:
+```
+function semver(major : Long, minor : Long, patch : Long) -> Semver {
+  { major : major, minor : minor, patch : patch }
+};
+```
+
+This would allow functions to typechecked in absence of a use-site, and allow for easier user specification of intent.
+It would also probably result in clearer type checker error messages.
+
+This introduces the following questions:
+
+1. Do we allow `type` declaration in policies, or just in schemas?
+2. Are type annotations enforced dynamically à la Contracts, or are they just ignored at runtime?
+  1. If they are dynamically enforced, that implies access to the schema to unfold type definitions. It also may introduce redundant type checking.
+3. Are type annotations required or optional?
+4. Will we have types to support generics, i.e., polymorphism?
```

Comment:
My answer:
1. If we ever have type annotations, they should not be in policies. We have so far had a tenet of a clear separation between policies, which are dynamically typed and have no type annotations, and schemas, which support static validation and enumerate/annotate with types.
2. Ignored, per above.
3. Optional, for maximal flexibility and simplicity
4. Do not support generics, but that's Ok when types are optional -- you can leave them off to get this